src/sidebar/components/MarkdownEditor.tsx

```diff
@@ -181,49 +189,96 @@ function ToolbarButton({
+  const [highlightedSuggestion, setHighlightedSuggestion] = useState(0);
+  const suggestions = useMemo(() => {
+    if (!atMentionsEnabled || activeMention === undefined) {
+      return [];
     }
 
-    const textarea = textareaRef.current!;
-    const listenerCollection = new ListenerCollection();
-    const checkForMentionAtCaret = () => {
-      const term = termBeforePosition(textarea.value, textarea.selectionStart);
-      setPopoverOpen(term.startsWith('@'));
-    };
-
-    // We listen for `keyup` to make sure the text in the textarea reflects the
-    // just-pressed key when we evaluate it
-    listenerCollection.add(textarea, 'keyup', e => {
-      // `Esc` key is used to close the popover. Do nothing and let users close
-      // it that way, even if the caret is in a mention
-      if (e.key !== 'Escape') {
-        checkForMentionAtCaret();
+    return usersWhoAnnotated
+      .filter(
+        u =>
+          // Match all users if the active mention is empty, which happens right
+          // after typing `@`
+          !activeMention ||
+          `${u.username} ${u.displayName ?? ''}`
+            .toLowerCase()
+            .match(activeMention.toLowerCase()),
+      )
+      .slice(0, 10);
+  }, [activeMention, atMentionsEnabled, usersWhoAnnotated]);
+
+  const checkForMentionAtCaret = useCallback(
+    (textarea: HTMLTextAreaElement) => {
+      if (!atMentionsEnabled) {
+        return;
       }
-    });
 
-    // When clicking the textarea it's possible the caret is moved "into" a
-    // mention, so we check if the popover should be opened
-    listenerCollection.add(textarea, 'click', checkForMentionAtCaret);
+      const term = termBeforePosition(textarea.value, textarea.selectionStart);
+      const isAtMention = term.startsWith('@');
 
-    return () => listenerCollection.removeAll();
-  }, [atMentionsEnabled, popoverOpen, textareaRef]);
+      setPopoverOpen(isAtMention);
+      setActiveMention(isAtMention ? term.substring(1) : undefined);
+
+      // Reset highlighted suggestion when closing the popover
+      if (!isAtMention) {
+        setHighlightedSuggestion(0);
+      }
+    },
+    [atMentionsEnabled],
+  );
+  const applySuggestion = useCallback(
+    (suggestion: UserItem) => {
+      const textarea = textareaRef.current!;
+      const { value } = textarea;
+      const { start, end } = getTermCoordinatesAtPosition(
+        value,
+        textarea.selectionStart,
+      );
+      const beforeMention = value.slice(0, start);
+      const beforeCaret = `${beforeMention}@${suggestion.username} `;
+      const afterMention = value.slice(end);
+
+      // Set textarea value directly, set new caret position and keep it focused.
+      textarea.value = `${beforeCaret}${afterMention}`;
+      textarea.selectionStart = beforeCaret.length;
+      textarea.selectionEnd = beforeCaret.length;
+      textarea.focus();
+      // Then update state to keep it in sync.
+      onEditText(textarea.value);
```

Comment:
This logic is similar to how we handle the toolbar commands. On a follow-up PR I will check if it's possible to treat applying a mention as a new type of command without making it too confusing.

Reply:
I'm not sure whether it will make sense to use the same code as the toolbar commands. However the main motivation behind the design of the text manipulation code for the toolbar commands is to make it easier to test them by splitting the logic out of the UI component. Similarly here I think it would be good to design the code in such a way that the behavior can be tested on its own without the surrounding `MarkdownEditor`.